core/src/main/java/org/apache/calcite/DataContexts.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite;
+
+import org.apache.calcite.adapter.java.JavaTypeFactory;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.linq4j.QueryProvider;
+import org.apache.calcite.schema.SchemaPlus;
+
+import com.google.common.collect.ImmutableMap;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+/** Utilities for {@link DataContext}. */
+public class DataContexts {
+  private DataContexts() {
+  }
+
+  /** Instance of {@link DataContext} that has no variables. */
+  public static final DataContext EMPTY = new EmptyDataContext();
+
+  /** Returns an instance of {@link DataContext} with the given map. */
+  public static DataContext of(Map<String, Object> map) {
```

Comment:
```suggestion
  public static DataContext of(Map<String, ? extends Object> map) {
```

Null values are not expected, however, `of(mapOf<String, String>(...))` would better be accepted